presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java

```diff
@@ -69,76 +61,97 @@ public KafkaSplitManager(
+            long[] offsets = findAllOffsets(leaderConsumer,  kafkaTableHandle.getTopicName(), part.partId(), startTs, endTs);
+            for (int i = offsets.length - 1; i > 0; i--) {
+                KafkaSplit split = new KafkaSplit(
+                        connectorId,
+                        kafkaTableHandle.getTopicName(),
+                        kafkaTableHandle.getKeyDataFormat(),
+                        kafkaTableHandle.getMessageDataFormat(),
+                        part.partId(),
+                        offsets[i],
+                        offsets[i - 1],
+                        partitionLeader,
+                        startTs,
+                        endTs);
+                splits.add(split);
+
+                long splitSize = (split.getEnd() - split.getStart()) / 1024 / 1024;
+                log.info("Split summarize: %s-%s (%sMB)", split.getStart(), split.getEnd(), splitSize);
+                estimatedTotalSize += splitSize;
             }
         }
 
-        return new FixedSplitSource(splits.build());
+        ImmutableList<ConnectorSplit> builtSplits = splits.build();
+        log.info("Built " + builtSplits.size() + " splits");
+
+        log.info("EstimatedTotalSize: %s", estimatedTotalSize);
+        return new FixedSplitSource(builtSplits);
     }
 
-    private static long[] findAllOffsets(SimpleConsumer consumer, String topicName, int partitionId)
+    private static long[] findAllOffsets(SimpleConsumer consumer, String topicName, int partitionId, Long startTs, Long endTs)
     {
-        TopicAndPartition topicAndPartition = new TopicAndPartition(topicName, partitionId);
-
-        // The API implies that this will always return all of the offsets. So it seems a partition can not have
-        // more than Integer.MAX_VALUE-1 segments.
-        //
-        // This also assumes that the lowest value returned will be the first segment available. So if segments have been dropped off, this value
-        // should not be 0.
-        PartitionOffsetRequestInfo partitionOffsetRequestInfo = new PartitionOffsetRequestInfo(kafka.api.OffsetRequest.LatestTime(), Integer.MAX_VALUE);
-        OffsetRequest offsetRequest = new OffsetRequest(ImmutableMap.of(topicAndPartition, partitionOffsetRequestInfo), kafka.api.OffsetRequest.CurrentVersion(), consumer.clientId());
-        OffsetResponse offsetResponse = consumer.getOffsetsBefore(offsetRequest);
-
-        if (offsetResponse.hasError()) {
-            short errorCode = offsetResponse.errorCode(topicName, partitionId);
-            log.warn("Offset response has error: %d", errorCode);
-            throw new PrestoException(KAFKA_SPLIT_ERROR, "could not fetch data from Kafka, error code is '" + errorCode + "'");
+        // startTs: start timestamp, or -2/null as earliest
+        // endTs: end timestamp, or -1/null as latest
+        startTs = startTs == null ? OffsetRequest.EarliestTime() : startTs;
+        endTs = endTs == null ? OffsetRequest.LatestTime() : endTs;
```

Comment:
will these ever be null? The attribute on `layoutHandle` is already checked for nullability - you could perhaps just use `long` here?